ee/codegen/src/project.ts

```diff
@@ -426,79 +404,84 @@ ${errors.slice(0, 3).map((err) => {
    */
   private getOrderedNodes(): WorkflowDataNode[] {
     const rawData = this.workflowVersionExecConfig.workflowRawData;
+    const nodes = rawData.nodes.filter(
+      (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
+    );
 
     // Edge case: Workflow init only has two nodes of ENTRYPOINT and TERMINAL with no edge between them
     if (rawData.edges.length === 0) {
-      return rawData.nodes.filter(
-        (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
-      );
+      return nodes;
     }
 
-    const nodesById = Object.fromEntries(
-      rawData.nodes.map((node) => [node.id, node])
-    );
+    const nodesById = Object.fromEntries(nodes.map((node) => [node.id, node]));
 
-    const edgesQueue = this.workflowContext.getEntrypointNodeEdges();
-    const edgesByPortId = this.workflowContext.getEdgesByPortId();
-    const processedEdges = new Set<WorkflowEdge>();
-    const processedNodeIds = new Set<string>();
+    // Create an adjacency list representation of the graph
+    const graph: Map<string, string[]> = new Map();
+    // Track in-degree (number of incoming edges) for each node
+    const inDegree: Map<string, number> = new Map();
```

Comment:
[q] Correct me if I'm wrong, but aren't we only using `inDegree` to decide which source nodes to start the algorithm at? If so, shouldn't we only need to calculate the initial source nodes?

Reply:
Oh! I missed this bit:

`if (currentDegree - 1 === 0) { ... }`

I see, we are only adding it to our array of nodes to process if all of it's parents are processed. Neat! I wonder if our graph generation algo would simplify using this same approach now that I think about it...

Reply:
[blocking] Wait, does this algorithm handle loops?

> A >> B >> C >> D >> B

Reply:
I think the answer is not really, bc it will just add them at the end as `unprocessedNodes`